core/shared/platform/zephyr/zephyr_file.c

```diff
@@ -0,0 +1,990 @@
+        // printf.
+        for (int i = 0; i < iovcnt; i++) {
+            if (iov[i].buf_len == 0)
+                continue;
+            os_printf("%s", (char *)iov[i].buf);
+            total_written += iov[i].buf_len;
+
+            // Clear the buffer after printing
+            memset(iov[i].buf, 0, iov[i].buf_len);
+        }
+        *nwritten = total_written;
+
+        return __WASI_ESUCCESS;
+    }
+
+    // Write data from each buffer
+    for (int i = 0; i < iovcnt; i++) {
+        ssize_t bytes_written =
+            fs_write(&ptr->file, iov[i].buf, iov[i].buf_len);
+        if (bytes_written < 0) {
+            return convert_errno(-bytes_written);
+        }
+
+        total_written += bytes_written;
+
+        // If we wrote less than we asked for, stop writing
+        if (bytes_written < iov[i].buf_len) {
+            break;
+        }
+    }
+
+    *nwritten = total_written;
+
+    return __WASI_ESUCCESS;
+}
+
+__wasi_errno_t
+os_fallocate(os_file_handle handle, __wasi_filesize_t offset,
+             __wasi_filesize_t length)
+{
+    return __WASI_ENOSYS;
+}
+
+__wasi_errno_t
+os_ftruncate(os_file_handle handle, __wasi_filesize_t size)
+{
+    struct zephyr_fs_desc *ptr = NULL;
+
+    GET_FILE_SYSTEM_DESCRIPTOR(handle->fd, ptr);
+
+    int rc = fs_truncate(&ptr->file, (off_t)size);
+    if (rc < 0) {
+        return convert_errno(-rc);
+    }
+
+    return __WASI_ESUCCESS;
+}
+
+__wasi_errno_t
+os_futimens(os_file_handle handle, __wasi_timestamp_t access_time,
+            __wasi_timestamp_t modification_time, __wasi_fstflags_t fstflags)
+{
+    return __WASI_ENOSYS;
+}
+
+__wasi_errno_t
+os_utimensat(os_file_handle handle, const char *path,
+             __wasi_timestamp_t access_time,
+             __wasi_timestamp_t modification_time, __wasi_fstflags_t fstflags,
+             __wasi_lookupflags_t lookup_flags)
+{
+    return __WASI_ENOSYS;
+}
+
+__wasi_errno_t
+os_readlinkat(os_file_handle handle, const char *path, char *buf,
+              size_t bufsize, size_t *nread)
+{
+    return __WASI_ENOSYS;
+}
+
+__wasi_errno_t
+os_linkat(os_file_handle from_handle, const char *from_path,
+          os_file_handle to_handle, const char *to_path,
+          __wasi_lookupflags_t lookup_flags)
+{
+    return __WASI_ENOSYS;
+}
+
+__wasi_errno_t
+os_symlinkat(const char *old_path, os_file_handle handle, const char *new_path)
+{
+    return __WASI_ENOSYS;
+}
+
+__wasi_errno_t
+os_mkdirat(os_file_handle handle, const char *path)
+{
+    struct zephyr_fs_desc *ptr = NULL;
+    int index, rc;
+    char abs_path[MAX_FILE_NAME + 1];
+
+    if (handle == NULL) {
+        return __WASI_EINVAL; // Or another appropriate error code
+    }
+
+    snprintf(abs_path, MAX_FILE_NAME, "%s/%s", prestat_dir, path);
+
+    rc = fs_mkdir(abs_path);
+    if (rc < 0) {
+        return convert_errno(-rc);
+    }
+
+    ptr = zephyr_fs_alloc_obj(true, abs_path, &index);
+    if (!ptr) {
+        return __WASI_EMFILE;
```

Comment:
should we remove the dir (e.g., fs_unlink) before return error?

Reply:
Yes we should. Fixed it.